# **CSI 4142 Assignment 2 - Part 1: Cleaning** ##
**Group 119:**\
Sami Hassan,
300169285\
Yassine Sami, 300146704

> # **Introduction**
The goal of this report is to analyze a dataset and create a "clean data checker" that will clean a dataset using 10 different tests. The dataset used in this first part of the assignment explores transactions at a café with several errors to be found. To use this notebook, you will go to each validity checker section numbered from 1-10 and use the second cell to select a test attribute from the list of possible attributes, then hitting run on the code block.

In [ ]:
# imports
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from datetime import datetime

# loading data set
dataset = pd.read_csv('https://raw.githubusercontent.com/sami-hssn/csi4142-a2/main/dirty_cafe_sales.csv')

> # **Dataset 1: Cafe Sales - Dirty Data for Cleaning Training**

## Dataset Description
The dataset used for the cleaning portion of Assignment 2 is called **"Cafe Sales - Dirty Data for Cleaning Training"**, published by Ahmed Mohamed on Kaggle. This dataset contains multiple transactions made at a café with several missing values, inconsistent data and errors made for data cleaning. This dataset contains **10,000** entries and **8** attributes.


The attributes outlined by the dataset:

*   **Transaction ID:** A unique identifier for each transaction.
*   **Item:** The name of the menu item purchased.
*   **Quantity:** The quantity of the item purchased.
*   **Price Per Unit:** The price of a single menu item.
*   **Total Spent:** The total spent on the transaction.
*   **Payment Method:** The method of payment used for the order.
*   **Location:** The location where the transaction occurred.
*   **Transaction Date:** The date of the transaction.


Some assumptions:

*   Each transaction can only have one unique menu item.
*   Total Spent = Quantity * Price Per Unit

The original dataset had to be modified so all 10 errors could be found and tested.. Ex: some values in Quantity were set to 0, some values in Total Spent set to negative values, a duplicate value was added.

> # **1. Data Type Errors**

### Data Type Check

This test will ensure data entered in the fields are of the correct type. For example, we will take columns that only accept numeric data: **Quantity**, **Price Per Unit** and **Total Spent**.

Note: The dataset values are all strings, need to convert from string to float (all ints can be floats).

In [ ]:
attributes = ['Quantity', 'Price Per Unit', 'Total Spent']

# attribute selection
test_attribute = 'Price Per Unit'

In [ ]:
# function to check if a value can be converted to int or float as dataset values are all strings
def is_numeric(value):
    # empty values are invalid
    if pd.isna(value) or value == "":
        return False
    try:
        # try converting to float first (since all ints can be floats)
        float(value)
        return True
    except (ValueError, TypeError):
        return False


if test_attribute in dataset.columns:
    # find invalid values (values that cannot be converted to int or float)
    invalid_data = dataset[~dataset[test_attribute].apply(is_numeric)]

    if invalid_data.empty:
        print(f"All values in {test_attribute} are numeric.")
    else:
        print(f"{len(invalid_data)} data type errors found in the {test_attribute} column.")
        # outputting which transaction id the invalid value occurs at
        for _, row in invalid_data.iterrows():
            print(f"Transaction ID: {row['Transaction ID']}, {test_attribute}: {row[test_attribute]}")
else:
    print(f"{test_attribute} is not a valid column in the CSV file.")

### Findings
There are **479** data type errors in the **'Quantity'** column, **533** data type errors in the **'Price Per Unit'** column, and **502** data type errors in the **'Total Spent'** column.




> # **2. Range Errors**

### Range Check

This test will ensure data entered in the fields falls within a given range. For example, we will take columns **Quantity**, and **Total Spent**.

**Quantity** must be **greater than 0**, cannot have Total Spent greater than 0 when no items are purchased.

**Total Spent** must be **greater than 0.0**. Assume no gift cards or store credit (not an option for payment method) can be used that would allow for Quantity > 0 and Total Spent = 0.0.


Note: The dataset values are all strings, need to convert from string to float (all ints can be floats).

In [ ]:
attributes = ['Quantity', 'Total Spent']

# attribute selection
test_attribute = 'Total Spent'

In [ ]:
# function to check if a value is numeric and greater than 0
def is_valid(value):
    # empty values are invalid
    if pd.isna(value) or value == "":
        return False
    try:
        # try converting to float first (since all ints can be floats)
        numeric_value = float(value)
        # check if value > 0
        return numeric_value > 0
    except (ValueError, TypeError):
        return False


if test_attribute in dataset.columns:
      # find invalid values (values that are not numeric, are empty, or are <= 0)
      invalid_data = dataset[~dataset[test_attribute].apply(is_valid)]

      if invalid_data.empty:
          print(f"All values in {test_attribute} are within the range.")
      else:
          # outputting which transaction id the invalid value occurs at
          print(f"{len(invalid_data)} range errors found in {test_attribute}.")
          for _, row in invalid_data.iterrows():
              print(f"Transaction ID: {row['Transaction ID']}, {test_attribute}: {row[test_attribute]}")
else:
    print(f"{test_attribute} is not a valid column in the CSV file.")

517 range errors found in Total Spent.
Transaction ID: TXN_4271903, Total Spent: ERROR
Transaction ID: TXN_7958992, Total Spent: nan
Transaction ID: TXN_8927252, Total Spent: ERROR
Transaction ID: TXN_6650263, Total Spent: UNKNOWN
Transaction ID: TXN_4987129, Total Spent: nan
Transaction ID: TXN_6289610, Total Spent: UNKNOWN
Transaction ID: TXN_8495063, Total Spent: UNKNOWN
Transaction ID: TXN_7623634, Total Spent: nan
Transaction ID: TXN_8989148, Total Spent: nan
Transaction ID: TXN_8562645, Total Spent: UNKNOWN
Transaction ID: TXN_5526852, Total Spent: nan
Transaction ID: TXN_3544789, Total Spent: ERROR
Transaction ID: TXN_5003018, Total Spent: UNKNOWN
Transaction ID: TXN_3229409, Total Spent: UNKNOWN
Transaction ID: TXN_6371987, Total Spent: ERROR
Transaction ID: TXN_5307411, Total Spent: ERROR
Transaction ID: TXN_8559167, Total Spent: UNKNOWN
Transaction ID: TXN_3093219, Total Spent: UNKNOWN
Transaction ID: TXN_5939055, Total Spent: UNKNOWN
Transaction ID: TXN_2726848, Total Spent:

### Findings
There are **495** range errors in the **'Quantity'** column, and **517** range errors in the **'Total Spent'** column.

***Transaction ID, Quantity:***\
Transaction ID: TXN_3353214, Quantity: 0\
Transaction ID: TXN_5516804, Quantity: nan\
Transaction ID: TXN_1351151, Quantity: ERROR

***Transaction ID, Total Spent:***\
Transaction ID: TXN_6627338, Total Spent: -15.0\
Transaction ID: TXN_9617421, Total Spent: nan\
Transaction ID: TXN_4522012, Total Spent: ERROR

> # **3. Format Errors**

### Format Check

This test will ensure data entered in field is in the proper format. For example, we will take the **'Transaction Date'** column.

The proper format for the column in the dataset is:
**YYYY-MM-DD**

In [ ]:
test_attribute = 'Transaction Date'

In [ ]:
# function to check if a date is in the correct format "yyyy-mm-dd"
def is_valid_date(date_str):
    # empty values are invalid
    if pd.isna(date_str) or date_str == "":
        return False
    try:
        # try to parse the date in the "yyyy-mm-dd" format
        datetime.strptime(date_str, "%Y-%m-%d")
        return True
    except ValueError:
        # parsing failed, invalid
        return False

if test_attribute in dataset.columns:
  # find invalid dates (not in yyyy-mm-dd format, empty, or ERROR)
  invalid_data = dataset[~dataset[test_attribute].apply(is_valid_date)]

  if invalid_data.empty:
      print("All date values are in the correct format 'yyyy-mm-dd'.")
  else:
      # outputting which transaction id the invalid value occurs at
      print(f"{len(invalid_data)} format errors found in {test_attribute}.")
      for _, row in invalid_data.iterrows():
          print(f"Transaction ID: {row['Transaction ID']}, {test_attribute}: {row[test_attribute]}")
else:
  print(f"{test_attribute} is not a valid column in the CSV file.")

### Findings
There are **473** format errors in the **'Transaction Date'** column.

***Transaction ID, Transaction Date:***\
Transaction ID: TXN_5003018, Transaction Date: UNKNOWN\
Transaction ID: TXN_9401522, Transaction Date: 11-16-2023\
Transaction ID: TXN_7550498, Transaction Date: ERROR

> # **4. Consistency Errors**


### Consistency Check

This test will ensure that prices of items are consistent across data entries. We will take a look at the **'Item"** column and will select an item from the menu.


In [ ]:
items = ['Coffee', 'Tea', 'Sandwich', 'Salad', 'Cake', 'Cookie', 'Smoothie', 'Juice']
prices = [2, 1.5, 4, 5, 3, 1, 4, 3]

item_price_map = dict(zip(items, prices))
test_item = 'Juice';

expected_price = item_price_map[test_item]

In [ ]:
if test_item in items:
  # filter rows for the test item and valid price values
  valid_rows = dataset[
      (dataset['Item'] == test_item) &
      (dataset['Price Per Unit'].apply(is_valid))
  ]

  # perform the consistency check on valid rows
  inconsistent_rows = valid_rows[
      valid_rows['Price Per Unit'].astype(float) != expected_price
  ]

  # display the inconsistent rows
  if inconsistent_rows.empty:
      print(f"No consistency errors found for '{test_item}'. All valid 'Price Per Unit' values match the expected price of {expected_price}.")
  else:
      print(f"Consistency errors found for '{test_item}'. Rows with incorrect 'Price Per Unit':")
      print(inconsistent_rows)
else:
  print(f"{test_item} is not a valid menu item.")

Consistency errors found for 'Juice'. Rows with incorrect 'Price Per Unit':
    Transaction ID   Item Quantity Price Per Unit Total Spent Payment Method  \
193    TXN_7412722  Juice        4            3.6        12.0           Cash   
241    TXN_8570890  Juice        4            3.8        12.0            NaN   

     Location Transaction Date  
193  In-store       2023-04-07  
241       NaN       2023-04-14  


### Findings

*   **1** consistency error in **Coffee**.
*   **0** consistency errors with **Tea**.
*   **0** consistency errors with **Sandwich**.
*   **0** consistency errors with **Salad**.
*   **0** consistency errors with **Cake**.
*   **1** consistency error with **Cookie**.
*   **0** consistency errors with **Tea**.
*   **1** consistency error with **Juice**.





> # **5. Uniqueness Errors**

### Uniqueness Check

This test will ensure data entered in the **'Transaction ID'** column is unique, as this field should have unique entries.

In [ ]:
test_attribute = 'Transaction ID'

In [ ]:
if test_attribute in dataset.columns:
  # find invalid dates (not in yyyy-mm-dd format, empty, or ERROR)
  duplicate_ids = dataset[dataset.duplicated(test_attribute, keep=False)]

  if duplicate_ids.empty:
      print(f"All values in {test_attribute} are unique.")
  else:
      # outputting which transaction id the invalid value occurs at
      print(f"{len(invalid_data)} uniqueness errors found in {test_attribute}.")
      print(duplicate_ids)
else:
  print(f"{test_attribute} is not a valid column in the CSV file.")

3961 uniqueness errors found in Transaction ID.
  Transaction ID    Item Quantity Price Per Unit Total Spent Payment Method  \
0    TXN_1961373  Coffee        2            2.0         4.0    Credit Card   
1    TXN_1961373  Coffee        2            2.0         4.0    Credit Card   

   Location Transaction Date  
0  Takeaway       2023-09-08  
1  Takeaway       2023-09-08  


### Findings
There is **1** uniqueness errors in the **'Transaction ID'** column.

> # **6. Presence Errors**

### Presence Check

This test will ensure that no fields in the dataset are left blank.

In [ ]:
attributes = ['Transaction ID', 'Item', 'Quantity', 'Price Per Unit', 'Total Spent', 'Payment Method', 'Location', 'Transaction Date']

# attribute selection
test_attribute = 'Transaction Date'

In [ ]:
if test_attribute in dataset.columns:
  # find nonempty values
  empty_data = dataset[dataset[test_attribute].isna()]

  if empty_data.empty:
      print(f"All values in {test_attribute} are filled.")
  else:
      # outputting which transaction id the invalid value occurs at
      print(f"{len(empty_data)} format errors found in {test_attribute}.")
      for _, row in empty_data.iterrows():
          print(f"Transaction ID: {row['Transaction ID']}, {test_attribute}: {row[test_attribute]}")
else:
  print(f"{test_attribute} is not a valid column in the CSV file.")


### Findings
* **0** presence errors in the **'Transaction ID'** column.
* **333** presence errors in the **'Item'** column.
* **138** presence errors in the **'Quantity'** column.
* **179** presence errors in the **'Price Per Unit'** column.
* **173** presence errors in the **'Total Spent'** column.
* **2579** presence errors in the **'Payment Method'** column.
* **3265** presence errors in the **'Location'** column.
* **159** presence errors in the **'Transaction Date'** column.

> # **7. Length Errors**

### Length Check

This test will ensure data entered in the **'Transaction ID'** must be 11 characters long. Each ID should be 3 letters, 1 underscore, and a 7 digit number, adding up to 11 characters.

In [ ]:
test_attribute = 'Transaction ID'

In [ ]:
if test_attribute in dataset.columns:
  # find IDs that are not 11 characters long
  invalid_data = dataset[dataset['Transaction ID'].str.len() != 11]

  if invalid_data.empty:
      print(f"All values in 'Transaction ID' are exactly 11 characters long.")
  else:
      # outputting which transaction id the invalid value occurs at
      print(f"{len(invalid_data)} length errors found in 'Transaction ID'.")
      print(invalid_data)
else:
  print(f"{test_attribute} is not a valid column in the CSV file.")

All values in 'Transaction ID' are exactly 11 characters long.


### Findings
There are **no** length errors in the **'Transaction ID'** column.

> # **8. Look-Up Errors**

### Look-Up Check

This test will ensure data entered in field is only from a set of limited values. For example, we will take the **'Item'**, **'Payment Method'**, and **'Location'** columns, which only have a certain number of acceptable values.


**Item:** [Coffee, Tea, Sandwich, Salad, Cake, Cookie, Smoothie, Juice]\
**Payment Method:** [Cash, Credit Card, Digital Wallet]\
**Location:** [In-store, Takeaway]

In [ ]:
attributes = ['Item', 'Payment Method', 'Location']

# attribute selection
test_attribute = 'Location'

In [ ]:
# define the acceptable values for each column
acceptable_values = {
    'Item': ['Coffee', 'Tea', 'Sandwich', 'Salad', 'Cake', 'Cookie', 'Smoothie', 'Juice'],
    'Payment Method': ['Cash', 'Credit Card', 'Digital Wallet'],
    'Location': ['In-store', 'Takeaway']
}

if test_attribute in dataset.columns:
    # find invalid values
    valid_values = acceptable_values[test_attribute]
    invalid_data = dataset[~dataset[test_attribute].isin(valid_values)]

    if invalid_data.empty:
      print(f"All values in '{test_attribute}' are valid.")
    else:
      # outputting which transaction id the invalid value occurs at
      print(f"{len(invalid_data)} look-up errors found in {test_attribute}.")
      for _, row in invalid_data.iterrows():
        print(f"Transaction ID: {row['Transaction ID']}, {test_attribute}: {row[test_attribute]}")
else:
    print(f"'{test_attribute}' is not a valid column in the CSV file.")

3961 look-up errors found in Location.
Transaction ID: TXN_7034554, Location: UNKNOWN
Transaction ID: TXN_2602893, Location: nan
Transaction ID: TXN_6699534, Location: UNKNOWN
Transaction ID: TXN_3765707, Location: nan
Transaction ID: TXN_8876618, Location: ERROR
Transaction ID: TXN_2616390, Location: nan
Transaction ID: TXN_7958992, Location: UNKNOWN
Transaction ID: TXN_8467949, Location: nan
Transaction ID: TXN_1736287, Location: nan
Transaction ID: TXN_8927252, Location: ERROR
Transaction ID: TXN_7710508, Location: nan
Transaction ID: TXN_9130559, Location: UNKNOWN
Transaction ID: TXN_6688524, Location: nan
Transaction ID: TXN_1491578, Location: nan
Transaction ID: TXN_5455792, Location: nan
Transaction ID: TXN_9499313, Location: nan
Transaction ID: TXN_8201146, Location: nan
Transaction ID: TXN_8230936, Location: ERROR
Transaction ID: TXN_8914892, Location: nan
Transaction ID: TXN_3363746, Location: nan
Transaction ID: TXN_9023317, Location: nan
Transaction ID: TXN_2537617, Locatio

### Findings
There are **969** look-up errors in the **'Item'** column, **3178** look-up errors in the **'Payment Method'** column, and **3961** look-up errors in the **'Location'** column.

> # **9. Exact Duplicate Errors**

### Exact Duplicate Check

This test will look for any **exact duplicates** in the dataset. An exact duplicate is defined as two entries with **identical values across all 8 attributes** in the dataset.

In [ ]:
duplicates = dataset.duplicated(keep = False)

dataset[duplicates]

Transaction ID    Item Quantity Price Per Unit Total Spent Payment Method  \
0    TXN_1961373  Coffee        2            2.0         4.0    Credit Card   
4    TXN_1961373  Coffee        2            2.0         4.0    Credit Card   

   Location Transaction Date  
0  Takeaway       2023-09-08  
4  Takeaway       2023-09-08

### Findings
There are two rows in the dataset, rows 0 and 4, that are exact duplicates. Even their unique identifier, the Transaction ID, is the same value.

> # **10. Near Duplicate Errors**

### Near Duplicate Check

This test will look for any near duplicates in the dataset. A near duplicate is defined as two different entries that may be duplicate because their values aside from their unique identifiers are identical.

In [ ]:
# Define the columns to check for near-duplicates (exclude 'Transaction ID')
columns_to_check = ['Item', 'Quantity', 'Price Per Unit', 'Total Spent', 'Payment Method', 'Location', 'Transaction Date']

# Find near-duplicates (rows where all specified columns are the same)
near_duplicates = dataset.duplicated(subset=columns_to_check, keep=False)

# Display the near-duplicate rows
if near_duplicates.any():
    print(f"Number of near-duplicate rows: {near_duplicates.sum()}\n")
    near_duplicates_grouped = dataset[near_duplicates].groupby(columns_to_check)
    for group_key, group_data in near_duplicates_grouped:
        print(f"Near-duplicate group:")
        print(f"{group_data} \n")
else:
    print("No near-duplicate rows found.")

Number of near-duplicate rows: 281

Near-duplicate group:
     Transaction ID  Item Quantity Price Per Unit Total Spent Payment Method  \
2359    TXN_4811887  Cake        2            3.0         6.0           Cash   
5119    TXN_4858908  Cake        2            3.0         6.0           Cash   

      Location Transaction Date  
2359  In-store       2023-02-11  
5119  In-store       2023-02-11   

Near-duplicate group:
     Transaction ID  Item Quantity Price Per Unit Total Spent Payment Method  \
4221    TXN_6421446  Cake        2            3.0         6.0           Cash   
6168    TXN_9931637  Cake        2            3.0         6.0           Cash   

      Location Transaction Date  
4221  Takeaway       2023-09-17  
6168  Takeaway       2023-09-17   

Near-duplicate group:
     Transaction ID  Item Quantity Price Per Unit Total Spent  Payment Method  \
2734    TXN_6360436  Cake        2            3.0         6.0  Digital Wallet   
9840    TXN_4959078  Cake        2            

### Findings
There are **281** near duplicate rows in the dataset.

> # **Conclusion**

The goal of this report was to clean a messy "dirty" data set using 10 error checking methods. The "clean data checker" was responsible for checking errors that pertained with data types, range, format, consistency, uniqueness, presence, length and look-ups. The cleaning process also looked for exact and near duplicates of entries in the data set. The dataset did not contain errors for all 10 tests so manual entries had to be added with "dirty" data to show the checker in action.



> # **References**

*   Data filtering - https://www.statology.org/pandas-filter-by-column-value-not-equal/
*   Applying function iteratively - https://www.geeksforgeeks.org/different-ways-to-iterate-over-rows-in-pandas-dataframe/
*   Dictionary items in consistency check - https://www.kaggle.com/code/hamelg/python-for-data-7-dictionaries-and-sets
*   String to date-time in format check - https://www.geeksforgeeks.org/python-datetime-strptime-function/
*   Finding duplicates - Caroline Barrière, CSI 4142 Course Notes: Week 4 - Data Cleaning, Part 1, Slide 27



